### ISSUE TITLE
Odd output with webpack-dev-server

### ISSUE THREAD

**Issue by andrewstart** [purtuga/esm-webpack-plugin#15]
While I'm sure that building a .js entry point with webpack-dev-server is probably an odd use case (testing in an external .html that pulls in the built library), I figured I'd report the bug I came across anyway.

I have a code bundle that just has a single default export. When building normally with `webpack`, I get the expected output:
```javascript
export default LIB['default'];
```
But when building with `webpack-dev-server` (no configuration changes), I get an unexpected output:
```javascript
export default LIB;
export { LIB };

export default LIB;
export { LIB };

export default LIB['default'];
```

**Comment by riovir** [comment]
**TL;DR: as a workaround, disable Hot Module Replacement on the WebpackDevServer.**

I've tracked the issue to [this block](https://github.com/purtuga/esm-webpack-plugin/blob/master/esm-webpack-plugin.js#L51-L53) 

```js
} else {
   exports += `export default ${libVar};\nexport { ${libVar} };\n`
}
```

It adds an export to the end of the current module. But when `MultiModule` is received, this code, and probably a bit more gets executed than necessary. When I logged out what modules are triggering this behavior, I found a familiar one: the Hot Module Replacement! I kindly instructed the dev server to cool down a bit:

```js
{
  devServer: {
    hot: false,
  }
}
```

The problematic chunk disappeared and my module started making sense. What would be nice from this plugin, is if it could detect "irrelevant modules", skipping over them.

**Comment by purtuga** [comment]
Hi @andrewstart + @riovir ,
Sorry I have not yet had a chance to look into this.  My guess here is that it has to do with it continuously building based on changes. I will look into in debug mode to understand it better and see if there is a way around it.

**Comment by purtuga** [comment]
Hi @andrewstart , @riovir - can you provide a quick repo that I can use to debug this?

**Comment by riovir** [comment]
On it...

**Comment by riovir** [comment]
Here you go: https://github.com/riovir/esm-webpack-plugin-i15-reprod

I've added some extra hoops, such as code splitting the bundled library to indicate it would still work if that pesky webpack-generated, extra HMR module wouldn't topple the plugin over.

If the codebase is too bulky, don't hesitate to ask for a simpler one!

**Comment by purtuga** [comment]
Thank you for that @riovir - it was a huge help. 
I have to say that I never envisioned using this plugin during development - especially with webpack devServer. That plugin adds lots of "things" to the code during hot reload. My plugin was more for when emitting a distribution build.

That being said, I may be able to introduce an option to the plugin that will allow you to "skip" processing certain bundled modules - like the ones from `webpack-dev-server`. I don't want to introduce specific code to handle that plugin webpack functionality, so I think providing a way for the user to "skip" some modules will work best.

There are two things going on that I can see with your provided example (there may be more issues as you work with a large real project):

1. The dev server hot reload functionality adds adds additional dependencies to the Entry Module - so although your webpack config has `entry.index: './src/index.js'`, this actually gets changed to include two additional files before this one, so it becomes a multi-Module entry point and thus the additional `export`'s on the initial built bundle. To address this, I will need to introduce a new option (a callback) - something like `skipModule(moduleName, module): boolean`.
To get around this problem in your sample project, I added this to line 34 of `esm-webpack-plugin.js`:
```javascript
    const moduleName = typeof module.nameForCondition === 'function' ? module.nameForCondition() : undefined;
    if (moduleName) {
        if (/[\\\/]webpack(-dev-server)?[\\\/]/.test(moduleName)) {
            return '';
        }
    }
```
2. The dev server also generates additional bundles as you change and save the source - these are also incorrectly being processed by my plugin. Good news is that I already have a way to Skip these - the `exclude()` option to the plugin, like:
```javascript
    	plugins: [
		new EsmWebpackPlugin({
			exclude(fileName) {
				// Exclude if:
				//	a. not a js file
				//  b. is a devServer.hot file
				return !/\.[cm]?js$/i.test(fileName) ||
					/\.hot-update\.js$/i.test(fileName);
			}
		}),
	],
```
You can see this change here: https://github.com/riovir/esm-webpack-plugin-i15-reprod/compare/master...purtuga:master

####  Re: issue 1 and new option
I'll see if I have some time tonight or tomorrow to introduce that options and push through a minor version update to the addin.

**Comment by purtuga** [comment]
@riovir - Ok 😄 . I added support for a new options called `skipModule`, but have not released it. Could you pull this plugin directly to your project from source and try it out?  If it all looks good to you, then I will cut a new minor release for it.

### Instructions:
Change your package.json and set this plugin `version` to be:

```
"devDependencies": {
    "@purtuga/esm-webpack-plugin": "purtuga/esm-webpack-plugin",
```
(this should work - its a been a while 🙂 . If not, then just install it with `npm i purtuga/esm-webpack-plugin`)

Then define the plugin in your webpack config as:

```javascript
new EsmWebpackPlugin({
    exclude(fileName) {
        // Exclude if:
        //  a. not a js file
        //  b. is a devServer.hot file
        return !/\.[cm]?js$/i.test(fileName) ||
            /\.hot-update\.js$/i.test(fileName);
    },
    skipModule(fileName, module) {
        return /[\\\/]webpack(-dev-server)?[\\\/]/.test(moduleName);
    }
})
```

**Comment by andrewstart** [comment]
It seems to work fine for me, the dev server runs fine and the normal build continues to build correctly.
I did have to correct your `skipModule` function though:
```javascript
skipModule(fileName, module) {
    return /[\\\/]webpack(-dev-server)?[\\\/]/.test(fileName);
}
```

**Comment by riovir** [comment]
I'll need some time to get back to checking the code out. But reading the comments, the idea seems to be working. I'll look into it next week. If that's too slow, please do go ahead and release the new version. 🙂

**Comment by purtuga** [comment]
Good to know its working as intended. I will try to release a new minor this weekend.

**Comment by purtuga** [comment]
Sorry for the delay.
I have published v1.3.0 to NPM - https://www.npmjs.com/package/@purtuga/esm-webpack-plugin - that includes this new options `skipModule`.
